x-pack/plugins/actions/server/action_service.ts

```diff
@@ -0,0 +1,154 @@
+import { SavedObjectReference } from './types';
+
+interface Action {
+  description: string;
+  actionTypeId: string;
+  actionTypeConfig: Record<string, any>;
+}
+
+interface FireOptions {
+  id: string;
+  params: Record<string, any>;
+  savedObjectsClient: SavedObjectsClient;
+}
+
+interface FindOptions {
+  perPage?: number;
+  page?: number;
+  search?: string;
+  defaultSearchOperator?: 'AND' | 'OR';
+  searchFields?: string[];
+  sortField?: string;
+  hasReference?: {
+    type: string;
+    id: string;
+  };
+  fields?: string[];
+}
+
+interface CreateOptions {
+  migrationVersion?: Record<string, string>;
+  references?: SavedObjectReference[];
+}
+
+export class ActionService {
+  private actionTypeService: ActionTypeService;
+  private encryptedSavedObjectsPlugin: EncryptedSavedObjectsPlugin;
+
+  constructor(
+    actionTypeService: ActionTypeService,
+    encryptedSavedObjectsPlugin: EncryptedSavedObjectsPlugin
+  ) {
+    this.actionTypeService = actionTypeService;
+    this.encryptedSavedObjectsPlugin = encryptedSavedObjectsPlugin;
+  }
+
+  /**
+   * Create an action
+   */
+  public async create(
+    savedObjectsClient: SavedObjectsClient,
+    data: Action,
+    options?: CreateOptions
+  ) {
+    const { actionTypeId } = data;
+    if (!this.actionTypeService.has(actionTypeId)) {
+      throw Boom.badRequest(`Action type "${actionTypeId}" is not registered.`);
+    }
+    this.actionTypeService.validateActionTypeConfig(actionTypeId, data.actionTypeConfig);
+    const actionWithSplitActionTypeConfig = this.moveEncryptedAttributesToSecrets(data);
+    return await savedObjectsClient.create('action', actionWithSplitActionTypeConfig, options);
+  }
+
+  /**
+   * Get an action
+   */
+  public async get(savedObjectsClient: SavedObjectsClient, id: string) {
+    return await savedObjectsClient.get('action', id);
+  }
+
+  /**
+   * Find actions
+   */
+  public async find(savedObjectsClient: SavedObjectsClient, options: FindOptions = {}) {
+    return await savedObjectsClient.find({
+      ...options,
+      type: 'action',
+    });
+  }
+
+  /**
+   * Delete action
+   */
+  public async delete(savedObjectsClient: SavedObjectsClient, id: string) {
+    return await savedObjectsClient.delete('action', id);
+  }
+
+  /**
+   * Update action
+   */
+  public async update(
+    savedObjectsClient: SavedObjectsClient,
+    id: string,
+    data: Action,
+    options: { version?: string; references?: SavedObjectReference[] } = {}
+  ) {
+    const { actionTypeId } = data;
+    if (!this.actionTypeService.has(actionTypeId)) {
+      throw Boom.badRequest(`Action type "${actionTypeId}" is not registered.`);
+    }
+    this.actionTypeService.validateActionTypeConfig(actionTypeId, data.actionTypeConfig);
+    const actionWithSplitActionTypeConfig = this.moveEncryptedAttributesToSecrets(data);
+    return await savedObjectsClient.update('action', id, actionWithSplitActionTypeConfig, options);
```

Comment:
Should we support partial updates? Otherwise we'll need the rest of the data from elasticsearch to pass validation.

Reply:
Yes